library/core/src/slice/index.rs

```diff
@@ -225,9 +225,10 @@ unsafe impl<T> SliceIndex<[T]> for usize {
 
     #[inline]
     unsafe fn get_unchecked(self, slice: *const [T]) -> *const T {
-        debug_assert_nounwind!(
-            self < slice.len(),
-            "slice::get_unchecked requires that the index is within the slice"
+        assert_unsafe_precondition!(
+            check_language_ub,
```

Comment:
all the other `slice::get_unchecked` family functions say that they're library UB, is this a mistake?

Reply:
It is not a mistake, this is the only unsafe slice accessor with an `assume` in it. I would like to add checks to the others, but doing so produces an icount regression and I can't find resultant optimizations that could justify it: https://github.com/rust-lang/rust/pull/120762. I did find a few such optimizations for the one assume that's in here.

Reply:
huh, optimizers are wild sometimes...

i suppose the line is a little fuzzy anyway, these methods are primarily (only?) used in the implementation of the reference based slice methods, which are causing language UB on invalid inputs.  it doesn't really matter though, since language UB checkers already know about this.

Reply:
The line isn't fuzzy at all actually. Without the `assume`, none of these functions execute language UB because they're all based on raw slices. Stacked Borrows rejects out-of-bounds `get_unchecked` but that's after these functions return and we reborrow the raw pointer into a reference. But that's equivalent to rejecting the `&Header` pattern, so there's no solid argument that these _must_ be language UB.

But someone on the community Discord just ran into this interesting optimization that LLVM refuses to do without the assumes: https://rust.godbolt.org/z/P31c58Ec1 I haven't dug into why LLVM can't do the optimization without the assumes here, but it's tantalizing.

Reply:
oh right, that pattern exists.

Reply:
So if someone adds `assume`s in the future, we have to be careful to adjust the preconditions?

Reply:
Hm. Actually I'm not so sure about things now.

Library UB checks are always on. If we bump these to Language UB, then they are ignored by const-eval/Miri. But checkers will still trip on the assume, albiet with a worse diagnostic.

Library UB escalated to language UB by adding an `assume` is an interesting case, because it almost doesn't matter what kind you call it. The only real difference is the details of the diagnostic, which are not great in both cases: We'd like for const-eval/Miri to be able to print details about the condition, but in the case of `assume(false)` and executing a library UB check, the interpreter is unaware.

Reply:
If the check says "library UB" but it's actually language UB, that's fairly harmless: it means some perf loss for CTFE and Miri and possibly worse error messages but no big deal.

If the check says "language UB" but it's actually library UB that's very unfortunate because the check is then entirely missed by Miri and so the backtrace where the error surfaced is further removed from the original cause.